From b7c9e5517af592c7fede0b24f8573ad234c083eb Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 13:13:07 -0400
Subject: [ethernet] i40evf: Fixed guest OS panic when removing vf driver

Message-id: <1409663591-32531-145-git-send-email-sassmann@redhat.com>
Patchwork-id: 90625
O-Subject: [RHEL7.1 PATCH 144/148] i40evf: Fixed guest OS panic when removing vf driver
Bugzilla: 1091127
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Serey Kong <serey.kong@intel.com>

Removing VF driver during device still in reset caused guest OS panic.

in the i40evf_remove(), we're trying to clean mac_filter_list which has
not been initialized since the device is still stuck at the reset.
The change is to initialize the filter_list before setting any task.

Change-ID: I8b59df7384416c7e6f2d264b598f447e1c2c92b0
Signed-off-by: Serey Kong <serey.kong@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 8bb1a540450c3dbd075491ea43772ac8a7ddec46)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40evf/i40evf_main.c b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
index e5679d8..ab15f4d 100644
--- a/drivers/net/ethernet/intel/i40evf/i40evf_main.c
+++ b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
@@ -2137,8 +2137,6 @@ static void i40evf_init_task(struct work_struct *work)
  ether_addr_copy(netdev->dev_addr, adapter->hw.mac.addr);
  ether_addr_copy(netdev->perm_addr, adapter->hw.mac.addr);
 
- INIT_LIST_HEAD(&adapter->mac_filter_list);
- INIT_LIST_HEAD(&adapter->vlan_filter_list);
  f = kzalloc(sizeof(*f), GFP_ATOMIC);
  if (NULL == f)
   goto err_sw_init;
@@ -2320,6 +2318,9 @@ static int i40evf_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
  hw->bus.device = PCI_SLOT(pdev->devfn);
  hw->bus.func = PCI_FUNC(pdev->devfn);
 
+ INIT_LIST_HEAD(&adapter->mac_filter_list);
+ INIT_LIST_HEAD(&adapter->vlan_filter_list);
+
  INIT_WORK(&adapter->reset_task, i40evf_reset_task);
  INIT_WORK(&adapter->adminq_task, i40evf_adminq_task);
  INIT_WORK(&adapter->watchdog_task, i40evf_watchdog_task);
-- 
1.7.1